src/backend/dx11/src/device.rs

```diff
@@ -1,983 +1,1322 @@
+use winapi::shared::{dxgiformat, dxgitype, winerror};
+use winapi::um::{d3d11, d3dcommon};
 
-impl core::mapping::Gate<R> for MappingGate {
-    unsafe fn set<T>(&self, index: usize, val: T) {
-        *(self.pointer as *mut T).offset(index as isize) = val;
-    }
+use wio::com::ComPtr;
 
-    unsafe fn slice<'a, 'b, T>(&'a self, len: usize) -> &'b [T] {
-        slice::from_raw_parts(self.pointer as *const T, len)
-    }
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::mem;
+use std::ops::Range;
+use std::ptr;
+use std::sync::Arc;
 
-    unsafe fn mut_slice<'a, 'b, T>(&'a self, len: usize) -> &'b mut [T] {
-        slice::from_raw_parts_mut(self.pointer as *mut T, len)
-    }
-}
+use {
+    Backend, Buffer, BufferView, CommandPool, ComputePipeline, DescriptorPool, DescriptorSetLayout,
+    Fence, Framebuffer, GraphicsPipeline, Image, ImageView, InternalBuffer, Memory, PipelineLayout,
+    QueryPool, RenderPass, Sampler, Semaphore, ShaderModule, Surface, Swapchain, UnboundBuffer,
+    UnboundImage, ViewInfo,
+};
 
-#[derive(Debug)]
-struct TextureParam {
-    levels: winapi::UINT,
-    format: winapi::DXGI_FORMAT,
-    bytes_per_texel: winapi::UINT,
-    bind: winapi::D3D11_BIND_FLAG,
-    usage: winapi::D3D11_USAGE,
-    cpu_access: winapi::D3D11_CPU_ACCESS_FLAG,
-}
+use {conv, internal, shader};
 
 pub struct Device {
-    device: ComPtr<winapi::ID3D11Device>,
-    share: Arc<Share>,
-    frame_handles: h::Manager<R>,
-    vs_cache: Map<u64, Vec<u8>>,
-    /// Create typed surface formats for the textures. This is useful for debugging
-    /// with PIX, since it doesn't understand typeless formats. This may also prevent
-    /// some valid views to be created because the typed formats can't be reinterpret.
-    use_texture_format_hint: bool,
-    sub_data_array: Vec<winapi::D3D11_SUBRESOURCE_DATA>,
-    feature_level: winapi::D3D_FEATURE_LEVEL,
+    pub(crate) device: ComPtr<d3d11::ID3D11Device>,
+    pub(crate) context: ComPtr<d3d11::ID3D11DeviceContext>,
+    pub(crate) memory_properties: hal::MemoryProperties,
+    pub(crate) internal: Arc<RefCell<internal::BufferImageCopy>>
 }
 
-impl Clone for Device {
-    fn clone(&self) -> Device {
-        Device::new(self.device.clone(), self.feature_level, self.share.clone())
-    }
-}
+unsafe impl Send for Device { }
+unsafe impl Sync for Device { }
 
 impl Device {
-    /// Create a new `Device`.
-    pub fn new(device: ComPtr<winapi::ID3D11Device>, feature_level: winapi::D3D_FEATURE_LEVEL, share: Arc<Share>) -> Device {
+    pub fn new(device: ComPtr<d3d11::ID3D11Device>, context: ComPtr<d3d11::ID3D11DeviceContext>, memory_properties: hal::MemoryProperties) -> Self {
         Device {
-            device: device,
-            share: share,
-            frame_handles: h::Manager::new(),
-            vs_cache: Map::new(),
-            use_texture_format_hint: false,
-            sub_data_array: Vec::new(),
-            feature_level: feature_level,
+            device: device.clone(),
```

Comment:
nit: could get away without a second `clone()`, or just re-order the fields to have plain `device,`